pkg/registry/apps/deployment/strategy.go

```diff
@@ -84,13 +86,18 @@ func (deploymentStrategy) PrepareForCreate(ctx context.Context, obj runtime.Obje
 func (deploymentStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
 	deployment := obj.(*apps.Deployment)
 	opts := pod.GetValidationOptionsFromPodTemplate(&deployment.Spec.Template, nil)
-	return validation.ValidateDeployment(deployment, opts)
+	return appsvalidation.ValidateDeployment(deployment, opts)
 }
 
 // WarningsOnCreate returns warnings for the creation of the given object.
 func (deploymentStrategy) WarningsOnCreate(ctx context.Context, obj runtime.Object) []string {
 	newDeployment := obj.(*apps.Deployment)
-	return pod.GetWarningsForPodTemplate(ctx, field.NewPath("spec", "template"), &newDeployment.Spec.Template, nil)
+	var warnings []string
+	if msgs := utilvalidation.IsDNS1123Label(newDeployment.Name); len(msgs) != 0 {
+		warnings = append(warnings, fmt.Sprintf("metadata.name: a DNS label is recommended: %v", msgs))
```

Comment:
general question: would it be more useful to indicate _why_ a DNS label is recommended in the warning rather than the validation failure message? as a user, I'd wonder what is breaking or not working well if I ignore this warning.

Reply:
The only really concrete WHY is hostnames.  Should we say that in the warning?  If I finish https://github.com/kubernetes/kubernetes/pull/106732 it will matter more

Reply:
I would find a message saying "pod hostnames may not work properly" helpful in deciding how much I cared about urgently fixing the warning on an existing thing

Reply:
I guess this is only a warning on create, which we'd hope they would encounter while developing their workload manifest, and we wouldn't spam people on updates to existing things... that's good

Reply:
How about: 

```
$ k create -f /tmp/bad.yaml 
Warning: metadata.name: this will become part of Pods' names and default hostnames, which can result in surprising behavior: a DNS label is recommended: [a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]
```

Reply:
that's better, but I think we can make it even clearer. If the primary difference between DNS label and DNS subdomain validation is that it can contain a `.`, can we simplify to:

```
if strings.Contains(newDeployment.Name, ".") {
   ...
   > `Warning: metadata.name: names containing '.' are not recommended; this name is used in pod names and default hostnames, and names containing '.' can result in surprising behavior`
}
```

Reply:
There are 2 primary criteria that are different: 1) dots; 2) max length.  We can check for (and warn on) those individually.  In fact, looking at `IsDNS1123Label()` it does check length independent of regex.  We could make that do something like:

```
errs := []
if too long {
  errs += "too long"
}
if valid subdomain but not valid label {
  errs += "no dots"
else 
  errs += regex
}
```

This would make nicer validation errors, too.  Result something like:

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must not contain dots]
```

or 

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must be no more than 63 characters]
```

or

```
Warning: metadata.name: this is used in Pod names and hostnames, which can result in surprising behavior: a DNS label is recommended: [must be no more than 63 characters, must not contain dots]
```

Or we could just do this in the warning logic and get the shorter errors you wrote

Reply:
I'm happy with any of those options... making the validation methods spit out nicer errors for everyone seems like an improvement, but one that will probably punish you by making you go change tests expecting specific nasty regex errors. Your call.

Reply:
```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go b/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
index e767092dd87..0b8a6cb354a 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go
@@ -191,7 +191,13 @@ func IsDNS1123Label(value string) []string {
                errs = append(errs, MaxLenError(DNS1123LabelMaxLength))
        }
        if !dns1123LabelRegexp.MatchString(value) {
-               errs = append(errs, RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
+               if dns1123SubdomainRegexp.MatchString(value) {
+                       // It was a valid subdomain and not a valid label.  Since we
+                       // already checked length, it must be dots.
+                       errs = append(errs, "must not contain dots")
+               } else {
+                       errs = append(errs, RegexError(dns1123LabelErrMsg, dns1123LabelFmt, "my-name", "123-abc"))
+               }
        }
        return errs
 }
```

Reply:
pushed, PTAL